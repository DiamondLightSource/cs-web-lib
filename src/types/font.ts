import { CSSProperties } from "react";

// assigned int values starting from zero
// also functions as a type
export enum FontStyle {
  Regular,
  Bold,
  Italic,
  BoldItalic
}

export class Font {
  // If no number is provided inherit from CSS.
  private size?: number;
  private style: FontStyle;
  private typeface: string;
  // TODO: name is currently unused
  private name?: string;

  public constructor(
    size?: number,
    style?: FontStyle,
    typeface?: string,
    name?: string
  ) {
    this.typeface = typeface ?? "Liberation sans";
    this.style = style ?? FontStyle.Regular;
    this.size = size;
    this.name = name;
  }

  /**
   * Returns a CSSProperties object containing all parameters defined by the
   * user and default properties for optional parameters not input by the user
   */
  public css(): CSSProperties {
    // Convert from px to REM by dividing pixels by default browser font size 16px
    // If a user has manually adjusted their browser font size, this might not be scaled
    // correctly, but given bob screens are pixel placed we shouldn't expect this to scale
    const fontSize = this.size ? `${this.size / 16}rem` : undefined;
    const fontWeight =
      this.style === FontStyle.Bold || this.style === FontStyle.BoldItalic
        ? "bold"
        : "normal";
    const fontStyle =
      this.style === FontStyle.Italic || this.style === FontStyle.BoldItalic
        ? "italic"
        : "normal";
    const style: CSSProperties = {
      // Fall back to sans-serif.
      fontFamily: `${this.typeface},sans-serif`,
      fontWeight: fontWeight,
      fontStyle: fontStyle
    };
    if (fontSize) {
      style.fontSize = fontSize;
    }
    return style;
  }
}
