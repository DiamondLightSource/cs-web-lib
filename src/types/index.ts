export { AbsolutePosition } from "./position";
export { RelativePosition } from "./position";
export { Color } from "./color";
export { PV } from "./pv";
export { Font, FontStyle } from "./font";
export { Border, BorderStyle } from "./border";
