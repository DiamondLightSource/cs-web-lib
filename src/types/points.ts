export class Point {
  public x: number;
  public y: number;

  public constructor(x: number, y: number) {
    this.x = x;
    this.y = y;
  }
}

export class Points {
  public values: Array<Point>;

  public constructor(values: Array<Point>) {
    this.values = values;
  }
}
