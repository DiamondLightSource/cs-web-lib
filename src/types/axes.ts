import { Color } from "./color";
import { Font, FontStyle } from "./font";

export class Axis {
  public index: number;
  public autoScale?: boolean;
  public autoScaleThreshold?: number;
  public axisColor?: Color;
  public axisTitle?: string;
  public showGrid?: boolean;
  public gridColor?: Color;
  public dashGridLine?: boolean;
  public timeFormat?: number;
  public scaleFormat?: string;
  public scaleFont?: Font;
  public titleFont?: Font;
  public visible?: boolean;
  public logScale?: boolean;
  public leftBottomSide?: boolean;
  public maximum?: number;
  public minimum?: number;
  public yAxis?: boolean;

  /**
   * Set default values for properties not yet
   * set, otherwise use set property. Uses same
   * default values as csstudio.opibuilder.xygraph.
   */
  public constructor(idx: number) {
    this.index = idx;
    this.autoScale = true;
    this.autoScaleThreshold = 0.95;
    this.axisColor = new Color("rgb(0, 0, 0");
    this.axisTitle = "";
    this.showGrid = false;
    this.gridColor = new Color("rgb(0, 0, 0");
    this.dashGridLine = true;
    this.timeFormat = 0;
    this.scaleFormat = "";
    this.scaleFont = new Font();
    this.titleFont = new Font(10, FontStyle.Bold);
    this.visible = true;
    this.logScale = false;
    this.leftBottomSide = true;
    this.minimum = 0;
    this.maximum = 100;
    // If axis has index 1 or above, assume y axis
    this.index > 0 ? (this.yAxis = true) : (this.yAxis = false);
  }
}

export class Axes {
  public count: number;
  public axisOptions: Axis[];

  public constructor(count: number, axes: Axis[]) {
    if (count !== axes.length) {
      throw new Error(
        `Count ${count} is not equal to number of axes ${axes.length}`
      );
    }
    this.count = count;
    this.axisOptions = axes;
  }
}
