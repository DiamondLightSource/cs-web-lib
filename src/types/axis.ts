import { Color } from "./color";
import { Font, FontStyle } from "./font";

export class Axis {
  public color: Color;
  public title: string;
  public showGrid: boolean;
  public visible: boolean;
  public logScale: boolean;
  public autoscale: boolean;
  public maximum: number;
  public minimum: number;
  public titleFont: Font;
  public scaleFont: Font;
  public onRight: boolean;

  /**
   * Set default values for properties not yet
   * set, otherwise use set property. Uses same
   * default values as csstudio.opibuilder.xygraph.
   */
  public constructor({
    xAxis = false,
    color = new Color("rgb(0, 0, 0"),
    title = "",
    showGrid = false,
    visible = true,
    logScale = false,
    autoscale = false,
    minimum = 0,
    maximum = 100,
    titleFont = new Font(FontStyle.Bold),
    scaleFont = new Font(),
    onRight = false
  } = {}) {
    this.color = color;
    this.title = title || (xAxis ? "X" : "Y");
    this.showGrid = showGrid;
    this.visible = visible;
    this.logScale = logScale;
    this.autoscale = autoscale;
    this.minimum = minimum;
    this.maximum = maximum;
    this.titleFont = titleFont;
    this.scaleFont = scaleFont;
    this.onRight = onRight;
  }
}

export type Axes = Axis[];
