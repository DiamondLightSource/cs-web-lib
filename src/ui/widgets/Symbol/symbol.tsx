import React, { useContext } from "react";

import { commonCss, Widget } from "../widget";
import { PVWidgetPropType, PVComponent } from "../widgetProps";
import {
  InferWidgetProps,
  BoolPropOpt,
  StringPropOpt,
  ColorPropOpt,
  FloatPropOpt,
  BorderPropOpt,
  ChoicePropOpt,
  FontPropOpt,
  ActionsPropType,
  StringArrayPropOpt
} from "../propTypes";
import { registerWidget } from "../register";
import { ImageComponent } from "../Image/image";
import { LabelComponent } from "../Label/label";
import { Color } from "../../../types/color";
import { executeActions, WidgetActions } from "../widgetActions";
import { MacroContext } from "../../../types/macros";
import { ExitFileContext, FileContext } from "../../../misc/fileContext";
import { DType } from "../../../types/dtypes";
import classes from "./symbol.module.css";

const SymbolProps = {
  imageFile: StringPropOpt,
  symbols: StringArrayPropOpt,
  alt: StringPropOpt,
  backgroundColor: ColorPropOpt,
  showBooleanLabel: BoolPropOpt,
  labelPosition: ChoicePropOpt([
    "top",
    "left",
    "center",
    "right",
    "bottom",
    "top left",
    "top right",
    "bottom left",
    "bottom right"
  ]),
  border: BorderPropOpt,
  rotation: FloatPropOpt,
  flipHorizontal: BoolPropOpt,
  flipVertical: BoolPropOpt,
  visible: BoolPropOpt,
  stretchToFit: BoolPropOpt,
  actions: ActionsPropType,
  font: FontPropOpt,
  initialIndex: FloatPropOpt,
  showIndex: BoolPropOpt,
  arrayIndex: FloatPropOpt,
  enabled: BoolPropOpt,
  fallbackSymbol: StringPropOpt,
  transparent: BoolPropOpt
};

export type SymbolComponentProps = InferWidgetProps<typeof SymbolProps> &
  PVComponent;

/**
 * This component combines the use of a svg with a label, and is used to replace
 * the MultistateMonitorWidget from CS-Studio
 * @param props
 */
export const SymbolComponent = (props: SymbolComponentProps): JSX.Element => {
  const {
    showIndex = false,
    arrayIndex = 0,
    initialIndex = 0,
    fallbackSymbol = "https://cs-web-symbol.diamond.ac.uk/catalogue/default.svg",
    transparent = true,
    backgroundColor = "white",
    showBooleanLabel = false,
    enabled = true
  } = props;
  const style = commonCss(props as any);
  // If symbols and not imagefile, we're in a bob file
  const isBob = props.symbols ? true : false;
  const symbols = props.symbols ? props.symbols : [];

  // Convert our value to an index, or use the initialIndex
  const index = convertValueToIndex(props.value, initialIndex, arrayIndex);

  const regex = / [0-9]\./;
  let imageFile = isBob ? symbols[index] : props.imageFile;
  // If no provided image file
  if (!imageFile) imageFile = fallbackSymbol;
  const intValue = DType.coerceDouble(props.value);
  if (!isNaN(intValue) && !isBob) {
    imageFile = imageFile.replace(regex, ` ${intValue.toFixed(0)}.`);
  }

  let alignItems = "center";
  let justifyContent = "center";
  switch (props.labelPosition) {
    case "top":
      alignItems = "flex-start";
      break;
    case "right":
      justifyContent = "flex-end";
      break;
    case "bottom":
      alignItems = "flex-end";
      break;
    case "left":
      justifyContent = "flex-start";
      break;
    case "top right":
      alignItems = "flex-start";
      justifyContent = "flex-end";
      break;
    case "bottom right":
      alignItems = "flex-end";
      justifyContent = "flex-end";
      break;
    case "bottom left":
      alignItems = "flex-end";
      justifyContent = "flex-start";
      break;
    case "top left":
      alignItems = "flex-start";
      justifyContent = "flex-start";
      break;
  }

  const files = useContext(FileContext);
  const exitContext = useContext(ExitFileContext);
  const parentMacros = useContext(MacroContext).macros;
  function onClick(event: React.MouseEvent<HTMLDivElement>): void {
    if (props.actions !== undefined && enabled) {
      executeActions(
        props.actions as WidgetActions,
        files,
        exitContext,
        parentMacros
      );
    }
  }

  // Define label appearance
  let labelDiv;
  if (isBob) labelDiv = generateIndexLabel(index, showIndex);

  // Note: I would've preferred to define the onClick on div that wraps
  // both sub-components, but replacing the fragment with a div, with the way
  // the image component is written causes many images to be of the incorrect size
  return (
    <>
      <ImageComponent
        {...props}
        imageFile={imageFile}
        onClick={onClick}
        stretchToFit={showBooleanLabel ? false : true}
        overflow={true}
      />
      {isBob ? (
        labelDiv
      ) : showBooleanLabel ? (
        <>
          <div
            className={classes.SymbolLabel}
            onClick={onClick}
            style={{
              ...style,
              backgroundColor: transparent
                ? "transparent"
                : backgroundColor.toString(),
              alignItems: alignItems,
              justifyContent: justifyContent
            }}
          >
            <div style={{ padding: "5%" }}>
              <LabelComponent
                {...props}
                textAlignV="bottom"
                backgroundColor={Color.TRANSPARENT}
                text={props.value?.getStringValue()}
              ></LabelComponent>
            </div>
          </div>
        </>
      ) : (
        <></>
      )}
    </>
  );
};

/**
 * Return a div element describing how the label should look
 */
function generateIndexLabel(index: number, showIndex: boolean): JSX.Element {
  if (!showIndex) return <></>;
  // Create span
  return (
    <div style={{ justifyContent: "center", alignContent: "center" }}>
      <span className={classes.IndexLabel}>
        <b>{index}</b>
      </span>
    </div>
  );
}

/**
 * Convert the input value into an index for symbols
 * @param value
 */
function convertValueToIndex(
  value: DType | undefined,
  initialIndex: number,
  arrayIndex: number
): number {
  // If no value, use initialIndex
  if (value === undefined) return initialIndex;
  // First we check if we have a string
  const isArray = value.getArrayValue()?.length !== undefined ? true : false;
  if (isArray) {
    // If is array, get index
    const arrayValue = DType.coerceArray(value);
    const idx = Number(arrayValue[arrayIndex]);
    return Math.floor(idx);
  } else {
    const intValue = DType.coerceDouble(value);
    if (!isNaN(intValue)) return Math.floor(intValue);
  }
  return initialIndex;
}

const SymbolWidgetProps = {
  ...SymbolProps,
  ...PVWidgetPropType
};

export const Symbol = (
  props: InferWidgetProps<typeof SymbolWidgetProps>
): JSX.Element => <Widget baseWidget={SymbolComponent} {...props} />;

registerWidget(Symbol, SymbolWidgetProps, "symbol");
