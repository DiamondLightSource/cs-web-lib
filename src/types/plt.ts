import { Axes, Axis } from "./axis";
import { Color } from "./color";
import { Font } from "./font";
import { Trace } from "./trace";

export class Plt {
  public title: string;
  public axes: Axes;
  public pvlist: Trace[];
  public backgroundColor: Color;
  public foregroundColor: Color;
  public scroll: boolean;
  public scrollStep: number;
  public updatePeriod: number;
  public start: string;
  public end: string;
  public showGrid: boolean;
  public titleFont: Font;
  public scaleFont: Font;
  public labelFont: Font;
  public legendFont: Font;

  /**
   * Set default values for properties not yet
   * set, otherwise use set property.
   */
  public constructor({
    title = "",
    axes = [new Axis()],
    pvlist = [new Trace()],
    background = Color.WHITE,
    foreground = Color.BLACK,
    scroll = true,
    grid = false,
    scrollStep = 5,
    updatePeriod = 1,
    titleFont = new Font(),
    labelFont = new Font(),
    legendFont = new Font(),
    scaleFont = new Font(),
    start = "1 minute",
    end = "now"
  } = {}) {
    this.backgroundColor = background;
    this.foregroundColor = foreground;
    this.title = title;
    this.scroll = scroll;
    this.scrollStep = scrollStep;
    this.axes = axes;
    this.pvlist = pvlist;
    this.updatePeriod = updatePeriod;
    this.showGrid = grid;
    this.start = start;
    this.end = end;
    this.titleFont = titleFont;
    this.scaleFont = scaleFont;
    this.labelFont = labelFont;
    this.legendFont = legendFont;
  }
}
