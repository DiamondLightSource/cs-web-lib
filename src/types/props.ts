import { Color } from "./color";
import { Font } from "./font";
import { MacroMap } from "./macros";
import { WidgetActions } from "../ui/widgets/widgetActions";
import { Border } from "./border";
import { Position } from "./position";
import { PV } from "./pv";
import { Trace } from "./trace";
import { Axes, Axis } from "./axis";
import { Points } from "./points";

export type GenericProp =
  | string
  | string[]
  | boolean
  | number
  | PV
  | Color
  | Font
  | Border
  | Position
  | Rule[]
  | MacroMap
  | WidgetActions
  | OpiFile
  | Trace[]
  | Axes
  | Axis
  | Points;

export interface Expression {
  boolExp: string;
  value: string;
  convertedValue?: GenericProp;
}

interface RulePV {
  pvName: PV;
  trigger: boolean;
}

export interface Rule {
  name: string;
  prop: string;
  outExp: boolean;
  pvs: RulePV[];
  expressions: Expression[];
}

export interface OpiFile {
  path: string;
  macros: MacroMap;
  defaultProtocol: string;
}
