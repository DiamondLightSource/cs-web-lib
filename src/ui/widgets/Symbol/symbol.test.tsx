import React from "react";
import { render, screen } from "@testing-library/react";
import { SymbolComponent } from "./symbol";
import { DType } from "../../../types/dtypes";

const fakeValue = new DType({ stringValue: "Fake value" });
const stringValue = new DType({ stringValue: "1.54" });
const arrayValue = new DType({ arrayValue: Float64Array.from([2, 0]) });

describe("<Symbol /> from .opi file", (): void => {
  test("label is not shown if showLabel is false", (): void => {
    const symbolProps = {
      showBooleanLabel: false,
      imageFile: "img 1.gif"
    };

    render(<SymbolComponent {...(symbolProps as any)} />);

    expect(screen.queryByText("Fake value")).not.toBeInTheDocument();
  });

  test("label is added", (): void => {
    const symbolProps = {
      showBooleanLabel: true,
      imageFile: "img 1.gif",
      pvData: [
        {
          value: fakeValue,
          connected: true,
          readonly: false,
          effectivePvName: "pv"
        }
      ]
    };
    render(<SymbolComponent {...(symbolProps as any)} />);

    expect(screen.getByText("Fake value")).toBeInTheDocument();
  });

  test("matches snapshot", (): void => {
    const symbolProps = {
      showBooleanLabel: true,
      imageFile: "img 1.gif",
      value: fakeValue
    };
    const { asFragment } = render(
      <SymbolComponent {...(symbolProps as any)} />
    );

    expect(asFragment()).toMatchSnapshot();
  });

  test("matches snapshot (with rotation)", (): void => {
    const symbolProps = {
      showBooleanLabel: false,
      imageFile: "img 1.gif",
      value: fakeValue,
      rotation: 45
    };

    const { asFragment } = render(
      <SymbolComponent {...(symbolProps as any)} />
    );

    expect(asFragment()).toMatchSnapshot();
  });
});

describe("<Symbol /> from .bob file", (): void => {
  test("index is not shown if showIndex is false", (): void => {
    const symbolProps = {
      symbols: ["img 1.gif"],
      value: new DType({ stringValue: "0" })
    };

    render(<SymbolComponent {...(symbolProps as any)} />);

    expect(screen.queryByText("0")).not.toBeInTheDocument();
  });

  test("index is added", (): void => {
    const symbolProps = {
      showIndex: true,
      symbols: ["img 1.gif", "img 2.png"],
      pvData: [
        {
          value: stringValue,
          connected: true,
          readonly: false,
          effectivePvName: "pv"
        }
      ]
    };
    render(<SymbolComponent {...(symbolProps as any)} />);

    expect(screen.getByText("1")).toBeInTheDocument();
  });

  test("use initialIndex if no props value provided", (): void => {
    const symbolProps = {
      showIndex: true,
      initialIndex: 2,
      symbols: ["img 1.gif", "img 2.png", "img 3.svg"],
      pvData: [
        {
          value: undefined,
          connected: true,
          readonly: false,
          effectivePvName: "pv"
        }
      ]
    };

    render(<SymbolComponent {...(symbolProps as any)} />);

    expect(screen.getByText("2")).toBeInTheDocument();
  });

  test("use arrayIndex to find index if value is an array", (): void => {
    const symbolProps = {
      arrayIndex: 0,
      showIndex: true,
      symbols: ["img 1.gif", "img 2.png", "img 3.svg"],
      pvData: [
        {
          value: arrayValue,
          connected: true,
          readonly: false,
          effectivePvName: "pv"
        }
      ]
    };
    render(<SymbolComponent {...(symbolProps as any)} />);

    expect(screen.getByText("2")).toBeInTheDocument();
  });

  test("matches snapshot (without index)", (): void => {
    const symbolProps = {
      symbols: ["img 1.gif"],
      value: new DType({ stringValue: "0" })
    };

    const { asFragment } = render(
      <SymbolComponent {...(symbolProps as any)} />
    );

    expect(asFragment()).toMatchSnapshot();
  });

  test("matches snapshot (with index)", (): void => {
    const symbolProps = {
      symbols: ["img 1.gif", "img 2.png", "img 3.svg"],
      value: new DType({ stringValue: "2" })
    };

    const { asFragment } = render(
      <SymbolComponent {...(symbolProps as any)} />
    );

    expect(asFragment()).toMatchSnapshot();
  });

  test("matches snapshot (using fallback symbol)", (): void => {
    const symbolProps = {
      symbols: ["img 1.gif"],
      value: new DType({ doubleValue: 1 })
    };
    const { asFragment } = render(
      <SymbolComponent {...(symbolProps as any)} />
    );

    expect(asFragment()).toMatchSnapshot();
  });

  test("matches snapshot (with rotation)", (): void => {
    const symbolProps = {
      symbols: ["img 1.gif"],
      value: new DType({ stringValue: "0" }),
      rotation: 45
    };

    const { asFragment } = render(
      <SymbolComponent {...(symbolProps as any)} />
    );

    expect(asFragment()).toMatchSnapshot();
  });
});
