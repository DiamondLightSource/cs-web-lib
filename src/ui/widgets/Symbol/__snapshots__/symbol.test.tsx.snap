// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<Symbol /> from .bob file > matches snapshot (using fallback symbol) 1`] = `
<DocumentFragment>
  <div
    style="overflow: visible; text-align: left; width: 100%; height: 100%;"
  >
    <img
      src="https://cs-web-symbol.diamond.ac.uk/catalogue/default.svg"
      style="width: 100%; height: 100%; display: block; transform: rotate(0deg) scaleX(1) scaleY(1); object-fit: fill; object-position: top left;"
    />
  </div>
</DocumentFragment>
`;

exports[`<Symbol /> from .bob file > matches snapshot (with index) 1`] = `
<DocumentFragment>
  <div
    style="overflow: visible; text-align: left; width: 100%; height: 100%;"
  >
    <img
      src="img 3.svg"
      style="width: 100%; height: 100%; display: block; transform: rotate(0deg) scaleX(1) scaleY(1); object-fit: fill; object-position: top left;"
    />
  </div>
</DocumentFragment>
`;

exports[`<Symbol /> from .bob file > matches snapshot (with rotation) 1`] = `
<DocumentFragment>
  <div
    style="overflow: visible; text-align: left; width: 100%; height: 100%;"
  >
    <img
      src="img 1.gif"
      style="width: 100%; height: 100%; display: block; transform: rotate(45deg) scaleX(1) scaleY(1); object-fit: fill; object-position: top left;"
    />
  </div>
</DocumentFragment>
`;

exports[`<Symbol /> from .bob file > matches snapshot (without index) 1`] = `
<DocumentFragment>
  <div
    style="overflow: visible; text-align: left; width: 100%; height: 100%;"
  >
    <img
      src="img 1.gif"
      style="width: 100%; height: 100%; display: block; transform: rotate(0deg) scaleX(1) scaleY(1); object-fit: fill; object-position: top left;"
    />
  </div>
</DocumentFragment>
`;

exports[`<Symbol /> from .opi file > matches snapshot (with rotation) 1`] = `
<DocumentFragment>
  <div
    style="overflow: visible; text-align: left; width: 100%; height: 100%;"
  >
    <img
      src="img 1.gif"
      style="width: 100%; height: 100%; display: block; transform: rotate(45deg) scaleX(1) scaleY(1); object-fit: fill; object-position: top left;"
    />
  </div>
</DocumentFragment>
`;

exports[`<Symbol /> from .opi file > matches snapshot 1`] = `
<DocumentFragment>
  <div
    style="overflow: visible; text-align: left; width: 100%; height: 100%;"
  >
    <img
      src="img 1.gif"
      style="width: 100%; height: 100%; display: block; transform: rotate(0deg) scaleX(1) scaleY(1); object-fit: fill; object-position: top left;"
    />
  </div>
  <div
    class="_SymbolLabel_7b7af1"
    style="background-color: transparent; align-items: center; justify-content: center;"
  >
    <div
      style="padding: 5%;"
    >
      <p
        class="MuiTypography-root MuiTypography-body1 css-io8o49-MuiTypography-root"
      >
        Fake value
      </p>
    </div>
  </div>
</DocumentFragment>
`;
